homeassistant/components/husqvarna_automower/coordinator.py

```diff
@@ -88,6 +88,30 @@ async def _async_update_data(self) -> MowerDictionary:
     @callback
     def handle_websocket_updates(self, ws_data: MowerDictionary) -> None:
         """Process websocket callbacks and write them to the DataUpdateCoordinator."""
+        self.hass.async_create_task(self._process_websocket_update(ws_data))
+
+    async def _process_websocket_update(self, ws_data: MowerDictionary) -> None:
+        """Handle incoming websocket update and update coordinator data."""
+        for data in ws_data.values():
+            existing_areas = data.work_areas or {}
+            for task in data.calendar.tasks:
+                work_area_id = task.work_area_id
+                if work_area_id is not None and work_area_id not in existing_areas:
+                    _LOGGER.debug(
+                        "New work area %s detected, refreshing data", work_area_id
+                    )
+                    # Important: async_refresh() is not used here because its update process
+                    # is asynchronous and may cause a race condition. Specifically, entities
+                    # may be created before the coordinator's data is updated, leading to
+                    # KeyErrors when accessing work area attributes.
+                    #
+                    # By explicitly polling, updating the coordinator, and then adding/removing
+                    # entities, we ensure consistent and complete data.
```

Comment:
Why does this happen tho

Reply:
I'm not sure. I tried hours to get it work. But if I use `async_refresh` or `_async_update_data` , then the data in the library is up to date, but self.data has still the old content.

Reply:
[this](https://github.com/home-assistant/core/pull/148358/commits/9ad6afd7300335b3360b12dceba8f965d24cace2) would work. But I'm not sure, if it's okay like this.